### DISCUSSION TITLE
아이템 15. 클래스와 멤버의 접근 권한을 최소화 하라

### DISCUSSION THREAD

**Discussion by JoisFe** [Study-2-Effective-Java/Effective-Java#41] [3. 과제] [answered]
# 아이템 15. 클래스와 멤버의 접근 권한을 최소화 하라

### 잘 설계된 컴포넌트
- 모든 내부 구현을 완벽히 숨김
- 구현과 API를 깔끔히 분리

즉 오직 API를 통해서만 다른 컴포넌트와 소통하며 서로의 내부 동작 방식에는 전혀 개의치 않는다.

### 장점
1. 시스템 개발 속도를 높임
2. 시스템 관리 비용을 낮춤
3. 정보 은닉 자체가 성능을 높여주지는 않지만 성능 최적화에 도움을 줌
4. 소프트웨어 재 사용성을 높임
5. 큰 시스템을 제작하는 난이도를 낮춰줌

## HOW) 어떻게 정보 은닉을 해야하는가 ?
우리는 자바에 대한 공부를 하고 있으니 자바에서는 어떻게 정보 은닉을 하는지 알아보자.

자바는 정보 은닉을 위한 다양한 장치를 제공함.
- 그 중 접근 제어 메커니즘은 클래스, 인터페이스, 멤버의 접근성(접근 허용 범위)을 명시
- 각 요소의 접근성은 그 요소가 선언된 위치와 접근 제한자로 정해짐

### 접근 제한자 종류
1. private : 멤버를 선언한 톱레벨 클래스에서만 접근 가능
2. package-private : 멤버가 소속된 패키지 안의 모든 클래스에서 접근할 수 있음
 > - 접근 제한자를 명시하지 않았을 때 적용되는 패키지 수준, 즉 default (단 interface의 멤버는 기본적으로 public)
3. protected : package-private의 접근 범위를 포함하며 이 멤버를 선언한 클래스의 하위 클래스에서도 접근할 수 있음
4. public : 모든 곳에서 접근 가능

이러한 접근제한자를 제대로 활용하는 원칙에 대해 알아보자

## 접근 제한자 활용 원칙
- 모든 클래스와 멤버의 접근성을 가능한 한 좁혀야 함. (소프트웨어가 올바로 동작하는 한 항상 가장 낮은 접근 수준을 부여해야 함)

### 1. 톱레벨 클래스(가장 바깥이라는 의미)와 인터페이스에 부여할 수 있는 접근 수준은 package-private, public 두가지 이다. 
 - 톱레벨 클래스나 인터페이스를 public으로 선언하면 공개 API가 됨
 - package-private로 선언하면 해당 패키지 안에서만 이용할 수 있음

패키지 외부에서 쓸 이유가 없다면 package-private로 선언하자

--> 그러면 API가 아닌 내부 구현이 되어 언제든 수정할 수 있음 (클라이언트에 아무런 피해 없이 수정, 교체, 제거 가능)
### 2. 한 클래스에서만 사용하는 package-private 톱레벨 클래스나 인터페이스는 이를 사용하는 클래스 안에 private static으로 중첩시키자 (아이템 24 참고)
- 톱레벨로 두면 같은 패키지의 모든 클래스가 접근할 수 있지만 private static으로 중첩시키면 바깥 클래스 하나에서만 접근할 수 있음

### 3. 클래스의 공개 API를 세심히 설계한 후, 그 외의 모든 멤버는 private로 만들자, 그런 다음 오직 같은 패키지의 다른 클래스가 접근해야 하는 멤버에 한하여 (private 제한자를 제거해) package-private로 풀어주자

- 권한을 풀어주는 일을 자주 하게 된다면 해당 시스템에서 컴포넌트를 더 분해해야 하는 것이 아닌지 다시 고민 해봐야함.
- private와 package-private 멤버는 모두 해당 클래스의 구현에 해당하므로 보통은 공개 API에 영향을 주지 않음 (단 Serializable을 구현한 클래스에서는 그 필드들도 의도치 않게 공개 API가 될 수도 있음 --> 아이템 86, 87 참고)

### 4. protected 멤버의 수는 적을 수록 좋다.
- public 클래스에서는 멤버의 접근 수준을 package-private에서 protected로 바꾸는 순간 그 멤버에 접근할 수 있는 대상 범위가 엄청나게 넓어진다.
- public 클래스의 protected 멤버는 공개 API 이므로 영원히 지원돼야 함.
- 내부 동작 방식을 API 문서에 적어 사용자에게 공개해야 할 수도 있음 (아이템 19 참고)

### 5. 리스코프 치환 원칙(#26 참고) 을 지키기 위해 클래스가 인터페이스가 정의한 모든 메서드를 public으로 선언해야하는 제약(예외)이 존재한다.
- 위 원칙을 제대로 지키지 못하는 경우가 발생 즉 멤버 접근성을 좁히지 못하게 방해하는 제약이 있음
- 상위 클래스의 메서드를 재정의할 때는 그 접근 수준을 상위 클래스에서보다 좁게 설정할 수 없음
- 이 제약은 상위 클래스의 인스턴스는 하위 클래스의 인스턴스로 대체해 사용할 수 있어야 하는 규칙인 리스코프 치환 원칙을 지키기 위해 필요
- 해당 규칙을 어길시 하위 클래스를 컴파일할 때 컴파일 오류가 발생
- 클래스가 인터페이스를 구현하는 건 이 규칙의 특별한 예
- 클래스는 인터페이스가 정의한 모든 메서드를 public으로 선언해야 함

### 6. 코드를 테스트하기 위한 목적으로 클래스, 인터페이스, 멤버의 접근 범위를 넓히진 말자 (적당한 수준까지는 괜찮음)

- 적당한 수준의 예로는 public 클래스의 private 멤버를 package-private 까지 풀어주는 것은 가능 (그 이상을 절대 안됨)
- 테스트 코드를 테스트 대상과 같은 패키지에 두면 package-private 요소에 접근 가능하기 때문
- 즉 테스트만을 위해 클래스, 인터페이스, 멤버를 공개 API로 만들어서는 안 된다.

### 7. public 클래스의 인스턴스 필드는 되도록 public이 아니어야 한다. (#39 참고)
- 필드가 가변 객체를 참조하거나 final이 아닌 인스턴스 필드를 public으로 선언하면 그 필드에 담을 수 없는 값을 제한할 힘을 잃게됨
- 해당 필드와 관련된 모든 것은 불변식으로 보장할 수 없다는 뜻
- 또한 필드가 수정될 때 (lock 획득 경우 등) 다른 작업을 할 수 없게 되므로 public 가변 필드를 갖는 클래스는 일반적으로 스레드에 안전하지 않음
- 필드에 final을 붙여 불변 객체를 참조하더라도 문제는 여전히 남음
- 내부 구현을 바꾸고 싶더라도 public 필드를 없애는 방식으로 refactoring 할 수 없게 되버림.

### 8. 7 의 예외가 존재 : 해당 클래스가 표현하는 추상 개념을 완성하는 데 꼭 필요한 구성요소로써의 상수라면 public static final 필드로 공개해도 됨.
- 7의 문제점은 정적 필드에서도 마찬가지
- 하지만 예외가 존재하는데 해당 클래스가 표현하는 추상 개념을 완성하는 데 꼭 필요한 구성요소로써의 상수인 경우 public static final 필드로 공개 가능
- 해당 상수의 표현 관례는 이름은 대문자 알파벳으로 쓰고 각 단어 사이에 밑줄(_)을 넣음 (아이템 68 참고)
- 또한 이러한 필드는 반드시 기본 타입 값이나 불변 객체를 참조 해야함 (#37 참고)
- 가변 객체를 참조한다면 final이 아닌 필드에 적용되는 모든 불이익이 그대로 적용되기 때문
- 이러한 경우 다른 객체를 참조하지 못하지만 참조된 객체 자체는 수정될 수 있으니 끔찍한 결과를 초래할 가능성이 존재함

### 9. 클래스에서 public static final 배열 필드를 두거나 이 필드를 반환하는 접근자 메서드를 제공해선 안된다.
- 이런 필드나 접근자를 제공하면 클라이언트에서 그 배열의 내용을 수정할 수 있게 됨.
``` java
public static final Thing[] VALUES = {};
```

위의 코드는 보안에 허점이 존재함.
어떤 IDE가 생성하는 접근자는 private 배열 필드의 참조를 반환하여 이 같은 문제를 똑같이 일으키니 주의해야 함!

### 해결책
두 가지 해결 방법이 존재한다.
1. 앞 코드의 public 배열을 private로 바꾸고 public 불변 리스트를 추가하는 것
``` java
public static final List<Thing> VALUES = {};

Collections.unmodifiableList(Arrays.asList(PRIVATE_VALUES));
// public static final List<Thing> VALUES = List.of(PRIVATE_VALUES);
// List.of() 또한 불변 리스트를 반환함
```

2. 배열을 private로 만들고 그 복사본을 반환하는 public 메서드를 추가하는 방법 (방어적 복사)
``` java
private static final Thing[] PRIVATE_VALUES = {};

public static final Thing[] values() {
    return PRIVATE_VALUES.clone();
}
```

- 클라이언트가 무엇을 원하느냐에 따라 위 둘 중 하나를 선택하면 됨
- 어느 반환 타입이 더 쓰기 편할지, 성능은 어느 쪽이 좋을지 고민해서 정하면 된다

## 모듈 시스템
JAVA 9 부터 모듈 시스템이라는 개념이 도입되면서 두 가지 암묵적 접근 수준이 추가됨

### 모듈 : 패키지 묶음
- 자신에 속한 패키지 중 공개 (export) 할 것들을 선언 (관례상 module-info.java 파일에 선언)

### 암묵적 접근 수준 : public, protected 수준의 효과가 모듈 내부로 한정
- protected 혹은 public 멤버라도 해당 패키지를 공개하지 않았다면 모듈 외부에서는 접근할 수 없다
- 물론 모듈 안에서는 exports로 선언했는지 여부에 아무런 영항도 받지 않음.
- 모듈 시스템을 활용하면 클래스 외부에 공개하지 않으면서도 같은 모듈을 이루는 패키지 사이에서는 자유롭게 공유 가능.

### 두가지 암묵적 접근 수준
- 숨겨진 패키지 안에 있는 public 클래스의 public 혹은 protected 멤버와 관련
- 해당 암묵적 접근 수준들은 각각 public 수준과 protected 수준과 같으나 그 효과가 모듈 내부로 한정되는 변종
- 이러한 형태로 공유해야하는 상황은 흔치 않음
- 만약 이런 상황이 발생하더라도 패키지들 사이에서 클래스들을 재배치 하면 대부분 해결됨
### 두가지 암묵적 접근 수준의 주의 사항
- 이전에 설명한 기존 접근 수준 (public ,package-private, protected, private)와 달리 새로운 두가지 암묵적 접근 수준은 상당히 주의해서 사용해야 함
- 모듈의 JAR 파일을 자신의 모듈 경로가 아닌 애플리케이션의 클래스패스 (classpath)에 두면 그 모듈 안의 모든 패키지는 마치 모듈이 없는 것처럼 행동
- 즉 모듈이 공개했는지 여부와 상관없이 public 클래스가 선언한 모든 public 혹은 protected 멤버를 모듈 밖에서도 접근할 수 있음

### 모듈 시스템의 활용
- 새로 등장한 해당 암묵적 접근 수준은 JDK에서 적극 활용 했음
- 자바 라이브러리에서 공개하지 않은 패키지들은 해당 모듈 밖에서는 절대 접근할 수 없음

### ETC
- 접근 보호 방식이 추가된 것 말고도 모듈은 여러 면에서 자바 프로그래밍에 영향을 끼쳤음
### 모듈의 장점을 활용하는 법
- 패키지들을 모듈 단위로 묶고 모듈 선언에 패키지들의 모든 의존성을 명시
- 소스 트리를 재배치하고 모듈 안으로부터 (모듈 시스템을 적용하지 않는) 일반 패키지로의 모든 접근에 특별한 조치를 취함

### 모듈 시스템의 현재
- JDK 외에도 모듈 개념이 널리 받아들여질지 아직 예측하기 이른 감이 있음.
- 따라서 꼭 필요한 경우가 아니라면... 당분간은 사용은 자제

### 모듈 시스템이라는 것도 있다

**Comment by corock** [comment, marked as answer]
접근 제한자 활용 원칙에서 5. ~~ 은 아직 *in progress* 로 보면 될까요?

**Reply by JoisFe** [reply]
아하 아직 정리중입니다. 빨리 정리하겠습니다. 😭

**Reply by corock** [reply]
### 접근 제한자 활용 원칙

> 톱레벨 클래스나 인터페이스를 `public` 으로 선언하면 공개 API 가 됨

*Spring Framework* 를 예로 들어보겠습니다.

일반적인 요청의 시작 지점이라고 할 수 있는 컨트롤러에 `public` 을 붙이면 `@RequestMapping` 인 메서드가 공개 API 가 된다고 생각합니다.
사실 컨트롤러는 내부적으로 API 를 호출하는 메서드에 `public` 접근 제한자를 붙이지 않고 `package-private` 만으로도 요청받을 수 있어요.
이 사례가 공개 API 와 그렇지 않은 API 의 적절한 예시인지는 잘 몰라서 @JoisFe 님을 포함하여 다른 분들 의견도 궁금합니다.

> `package-private` 에서 `protected` 로 바꾸는 순간 그 멤버에 접근할 수 있는 대상 범위가 엄청나게 넓어진다.

자바를 처음 학습할 때, 그리고 객체지향 프로그래밍의 4가지 특징 중 하나인 상속을 한창 접하고 있을 때 고민했던 점이 있습니다.
접근 제한자로 `protected` 를 만들어 놓았으니 상속 관계를 맺는 하위 클래스에는 `protected` 를 적극 사용하면 좋겠다고 생각했는데,
생각보다 `protected` 를 잘 사용하지 않는다고 느꼈고, 그 당시에는 명확한 이유를 찾지 못했었는데 식견이 넓어지는 시간이 되었네요 👍

**Reply by chikeem90** [reply]
> 일반적인 요청의 시작 지점이라고 할 수 있는 컨트롤러에 public 을 붙이면 @RequestMapping 인 메서드가 공개 API 가 된다고 생각합니다.
> 사실 컨트롤러는 내부적으로 API 를 호출하는 메서드에 public 접근 제한자를 붙이지 않고 package-private 만으로도 요청받을 수 있어요.
> 이 사례가 공개 API 와 그렇지 않은 API 의 적절한 예시인지는 잘 몰라서 @JoisFe 님을 포함하여 다른 분들 의견도 궁금합니다.

음,, 책에서 얘기하는 공개 API의 의미는 해당 기능을 외부 개발자에게 오픈하냐 안하냐의 의미인 거 같아요.
예를 들어 java에서 기본으로 제공하는 Comparable 처럼 public으로 오픈해두면 개발자가 가져다가 구현을 해서 compareTo를 마음대로 구현해서 사용할 수 있는 것이고,, 클래스를 예로 들면 ArrayList 같은 것도 public으로 오픈해두었기 때문에 우리가 직접 ArrayList 인스턴스를 만들고 public 메서드인 add도 사용할 수 있는거잖아요 ~? 
API의 정의를 웹 API 한정이 아니라 제공하는 기능 단위로 이해하면 될 거 같구요,, 그렇기 때문에 톱레벨의 클래스나 인터페이스를 public으로 선언하면 공개 API가 된다라고 할 수 있을 거 같네요.
(이거랑 별개로,, spring mvc로 개발할 때,, 접근 제한자를 public이냐 package-private이냐에 따라 공개된 API인지 내부 API인지 구분해서 개발하시는걸까요? 이 부분은 저도 정확하게 기억은 안나는데 스프링 내부 구현 방식 때문에 public으로 쓰는 것이 권장되고,, 접근제한자로 외부 공개 API이냐 내부 API이냐를 처리하는 건 아니었던 거 같아서요)

**Reply by JoisFe** [reply]
스프링 Controller의 핸들러 메서드 경우 package-private를 쓰지 않는 가장 큰 이유로는
해당 메서드를 트랜잭션으로 만들거나 다른 방식으로 proxy 해야 하는 경우 public 이어야 하기 떄문이라고 합니다.

public으로 써야하는 이유, 그리고 public으로 하지 않아 문제가 발생한 케이스 등의 좋은 사례가 있어 링크를 남기니 참고하시면 좋을 듯 해요!!

[public으로 써야 하는 이유](https://stackoverflow.com/questions/38262666/springmvc-controller-public-or-default-protected)
[public으로 하지 않아 문제가 발생한 케이스](https://stackoverflow.com/questions/46684838/private-method-with-springmvc)
[public으로 하지 않아 문제가 발생한 케이스2](https://jira.spring.io/projects/SPR/issues/SPR-16878?filter=allissues)

**Comment by coalong** [comment]
정리 감사합니다!

> 테스트 코드를 테스트 대상과 같은 패키지에 두면 package-private 요소에 접근 가능하기 때문
 
이 뜻이 테스트 코드와 테스트 대상 코드를 같은 패키지에 두고 작성하라는 뜻이겠죠? 실 개발할 때 위와 같은 방식으로 package-private 까지 접근 제한자를 열어두고  테스트 코드를 작성하는지 궁금합니다.!



> 암묵적 접근 수준 : public, protected 수준의 효과가 모듈 내부로 한정

> 모듈 시스템의 현재
JDK 외에도 모듈 개념이 널리 받아들여질지 아직 예측하기 이른 감이 있음.
따라서 꼭 필요한 경우가 아니라면... 당분간은 사용은 자제   

모듈 시스템과 암묵적 접근 수준은 저에게 생소한 개념이라 이번 기회에 알게 되었습니다.  조금 더 찾아봐야겠지만
JDK 9가 릴리즈되고 이펙티브 자바 책 3/E 이 17년도에 나왔는데, 혹시 현재의 모듈 시스템은 어떤지 궁금하네요..

**Reply by JoisFe** [reply]
### Q) : 이 뜻이 테스트 코드와 테스트 대상 코드를 같은 패키지에 두고 작성하라는 뜻이겠죠? 실 개발할 때 위와 같은 방식으로 package-private 까지 접근 제한자를 열어두고 테스트 코드를 작성하는지 궁금합니다.!

### A) : 
넵 package-private 접근 제한자 까지 범위를 넓히면 같은 패키지에 두면 테스트 코드는 작성할 수 있으니까 이해하신 대로 맞습니다.

사실 private 접근 지정자에 대한 부분에 무조건 접근하지 못하지는 않습니다.
왜냐하면 Java Reflection API가 있기 떄문이죠. 
예를 들어 private 메서드가 존재한다고 합시다.
리플렉션을 이용하면 메서드 코드를 메타정보로 추상화된 Method를 얻어낼 수 있으며 직접 호출 또한 가능하다고 합니다. 따라서 리플랙션을 이용해 테스트 하는 경우도 존재합니다.

다만 private로 걸어둔 메서드를 리플렉션을 통해 접근한다??
애초에 private로 메서드를 걸어둔 것은 클라이언트 입장에서 해당 메서드 내부를 전혀 알 필요가 없다는 의미 입니다. 즉 기능만 잘 유지된다면 내부 코드를 부담없이 바꿀 수 있죠..
하지만 해당 테스트 코드는 이미 private 메서드에 결합도가 높고 private 메서드의 내부 코드가 수정되었을 시 테스트가 깨질 가능성이 높습니다.

테스트 코드도 코드인만큼 유지 보수가 좋아야 겠죠.. 
즉 리플렉션을 이용한 테스트 코드는 지양 해야 합니다

그럼 도대체 어떻게 해야할지 저 또한 고민이 많았고 사실 Reflection을 이용해서 해당 코드를 테스트 하거나 (스프링 에서 ReflectionTestUtils 를 사용 하였음..) private 코드는 테스트 코드를 건너 뛰고 통합 테스트에서 괜찮다면 괜찮겠지 라는 핑계로 미루었습니다.

JUnit5에서 @ParameterizedTest라는 방법이 있다고는 하는데 해당 부분은 아직 제가 잘 모르는 분야라 대답하긴 힘들 것 같고 같이 알아볼 수 있으면 좋을 것 같습니다.

private 메서드를 테스트 해야 하는 상황이 설계의 문제가 있다는 의견 또한 존재하여 이 부분에 대해서도 같이 토론해 봤으면 좋겠습니다.
<br>

### Q) : JDK 9가 릴리즈되고 이펙티브 자바 책 3/E 이 17년도에 나왔는데, 혹시 현재의 모듈 시스템은 어떤지 궁금하네요..
### A) : 
저 또한 모듈에 대한 개념은 매우 생소하여 제대로 된 응답을 하지 못하겠습니다.
책에서는 아직 이른감이 있으니 당분간 사용은 자제하라는 말이 있었지만 effective java 책이 나온지 좀 되었다보니 저도 해당 정보에 대해 좀 더 알아보기 위해 찾아보았는데 쉽게 찾아지진 않네요...
다른 분들과 함께 좀 더 알아보면 좋을 것 같습니다.

다만 경험으로는 intelliJ IDE로 자바 프로젝트 혹은 스프링 프로젝트에서 Project Structure에 가 보면 module을 확인할 수 있습니다. 이 module이 아마 책에서 설명하는 module로 이해됩니다. 

![image](https://user-images.githubusercontent.com/90208100/210260948-7913ac1c-c46a-466c-84ed-0b5a440d0ed7.png)

@coalong 님 또한 QueryDSL을 사용하신 걸로 알고 있는데
![image](https://user-images.githubusercontent.com/90208100/210261115-4645e8d8-fde5-43cd-b76e-64788c22f2ce.png)

이 에러 기억 나십니까??

maven을 build tool로 사용 시 maven clean, compile, package 모든 짓을 다해봐도 이러한 에러가 잡히지 않을 떄가 있습니다.

![image](https://user-images.githubusercontent.com/90208100/210261303-4ef8c873-1c37-462e-b219-bb7a7d853a5c.png)
target의 generated-sources에 QEntity들이 잡히더라도 안잡히는 경우에
Project Structure에서 module에서 generated-sources 디렉터리를 Sources로 마크해주면 해결 되었던 경험이 있습니다.
![image](https://user-images.githubusercontent.com/90208100/210261397-a5bd4b47-0542-45bf-9c35-4ea20b213ecb.png)
![image](https://user-images.githubusercontent.com/90208100/210261416-4a8a9523-e34e-471e-b0ed-015b387784b7.png)

이러한 설정들이 모듈 시스템의 일부가 아닌지... (진짜 몰라서 옛 경험을 가져왔습니다. ) 왜냐하면 모듈이 패키지의 모음이다 보니까 패키지 전체를 한번에 설정하는 부분이 모듈을 건드리는 게 아닌가?? 이 정도만 생각해았습니다.
모듈 시스템에 대해 경험해 보셨거나 조금 더 아시는 스터디원 있으면 알려주시면 감사할 것 같습니다. ㅎㅎ


좋은 답변이 되었는지 잘 모르겠습니다. ㅎㅎ 질문 주셔서 감사하고 제 대답에 이상한 점이 있거나 다른 질문 있으시면 comment 남겨 주세요.

**Reply by chikeem90** [reply]
> 그럼 도대체 어떻게 해야할지 저 또한 고민이 많았고 사실 Reflection을 이용해서 해당 코드를 테스트 하거나 (스프링 에서 ReflectionTestUtils 를 사용 하였음..) private 코드는 테스트 코드를 건너 뛰고 통합 테스트에서 괜찮다면 괜찮겠지 라는 핑계로 미루었습니다.

이거는 재철님이 맞게 하고 계신 거 같아요,,
먼저 얘기주신대로 private이라는 거는 내부적으로만 사용되는 메서드이고, 해당 메서드를 사용하는 public 메서드들이 있을텐데 그 public 메서드를 테스트하면서 자연스럽게 private 메서드의 테스트가 되는 방향으로 처리하는게 제일 좋은 방식이었던 거 같아서요.
아니면 책에서 나온대로 package-private까지만 열어서 테스트코드도 동일한 패키지에 넣고 테스트하는 방법도 괜찮은 거 같네요.

**Reply by chikeem90** [reply]
> JUnit5에서 @ParameterizedTest라는 방법이 있다고는 하는데 해당 부분은 아직 제가 잘 모르는 분야라 대답하긴 힘들 것 같고 같이 알아볼 수 있으면 좋을 것 같습니다.

이거는 private 메서드를 테스트하는 방식은 아니고, 한 테스트로 여러 파라미터를 테스트할 수 있도록 JUnit에서 제공하는 기능입니다.
https://www.baeldung.com/parameterized-tests-junit-5
이거 참고해보세요.

```java
@ParameterizedTest
@ValueSource(ints = {1, 3, 5, -3, 15, Integer.MAX_VALUE}) // six numbers
void isOdd_ShouldReturnTrueForOddNumbers(int number) {
    assertTrue(Numbers.isOdd(number));
}
```

이런 식으로 쓰면 해당 테스트가 저 ValueSource에 정의된 파라미터에 대해 각각 테스트하게 됩니다.

**Reply by JoisFe** [reply]
올려주셔서 감사합니다 ㅎㅎ

**Reply by chikeem90** [reply]
https://www.jetbrains.com/help/idea/creating-and-managing-modules.html#modules-idea-java

**Comment by JoisFe** [comment]
@coalong  아 그리고 모듈 이야기가 나와서 생각해보니 스프링에서 maven, gradle 빌드 툴로 멀티 모듈 프로젝트 만드신 경험이 있는 것으로 압니다. 그 부분도 기억나신다면 정리해주시면 감사하겠습니다. .ㅎㅎ^^

**Reply by JoisFe** [reply]
그 모듈이 그 모듈이 아닌 것 같습니다. ㅎㅎ
